NWN.Permissions/src/ChatCommands/UserAddPermissionCommand.cs

```diff
@@ -0,0 +1,66 @@
+using System;
+using System.Collections.Generic;
+using Anvil.API;
+using Anvil.Services;
+using Jorteck.ChatTools;
+
+namespace Jorteck.Permissions
+{
+  [ServiceBinding(typeof(IChatCommand))]
+  internal class UserAddPermissionCommand : IChatCommand
+  {
+    private readonly ConfigService configService;
+
+    public string Command => configService.GetFullChatCommand("user addpermission");
+    public string[] Aliases => null;
+
+    public Dictionary<string, object> UserData { get; } = new Dictionary<string, object>
+    {
+      [PermissionConstants.ChatUserDataKey] = PermissionConstants.UserAddPermission,
+    };
+
+    public int? ArgCount => 1;
+    public string Description => "Grants a permission to a user.";
+
+    public CommandUsage[] Usages { get; } =
+    {
+      new CommandUsage("<permission_name>", "Grant the specified permission to the target user."),
+    };
+
+    public UserAddPermissionCommand(ConfigService configService)
+    {
+      this.configService = configService;
+    }
+
+    public void ProcessCommand(NwPlayer caller, IReadOnlyList<string> args)
+    {
+      string permission = args[0];
+      caller.EnterTargetMode(eventData => eventData.ProcessOnValidPCTargetBy(caller, AddUserPermissionToTarget(permission)));
```

Comment:
This felt like a convenient way of running actions on targets without having to duplicate the validation logic everywhere.

Reply:
++, I like this pattern :)

One thing which would also resolve the nesting issue in `AddUserPermissionToTarget` is to bump the extension to extend `NwPlayer` instead:

```cs
internal sealed class PlayerTargetPlayerEvent
{
  public NwPlayer Caller { get; }
  public NwPlayer Target { get; }

  public PlayerTargetPlayerEvent(NwPlayer caller, NwPlayer target)
  {
    Caller = caller;
    Target = target;
  }
}

internal static void EnterPlayerTargetMode(this NwPlayer player, Action<PlayerTargetPlayerEvent> callback, MouseCursor cursorType = MouseCursor.Magic, MouseCursor badTargetCursor = MouseCursor.NoMagic)
{
  void TargetHandler(ModuleEvents.OnPlayerTarget eventData)
  {
    if (eventData.IsCancelled)
    {
      return;
    }

    if (eventData.TargetObject is NwCreature targetCreature && targetCreature.IsLoginPlayerCharacter(out NwPlayer targetPlayer))
    {
      callback?.Invoke(new PlayerTargetPlayerEvent(eventData.Player, targetPlayer));
    }
    else
    {
      eventData.Player.SendErrorMessage("Target must be a player character.");
    }
  }

  player.EnterTargetMode(TargetHandler, ObjectTypes.Creature, cursorType, badTargetCursor);
}
```

Reply:
adde1b3c2a002ec0967304a1c3daee748727740d

I chose to pass `NwPlayerExtensions.PlayerTargetPlayerEvent` to the functions in order to not get the target and caller mixed up, at the cost of some boilerplate when navigating the selection properties. What do you think?